add pcrsys to predict PCR values

bcressey opened the pull request:
**Issue number:**
Related: https://github.com/bottlerocket-os/bottlerocket/issues/4680

**Description of changes:**
Add the `pcrsys` tool to predict PCR values at build time.

**Testing done:**
#### Notes
- PCR 5 has multiple valid values due to GPT partition table variations
- PCR 11 has multiple valid values representing different boot stages

#### x86_64-aws-ecs-3-fips
- **Status**: PASS
- **Disk prediction**: PASS
- **AMI prediction**: PASS
- **AMI ID**: ami-06dde408a20fb8183
- **Mismatched PCRs**: None

| PCR | Predicted | Actual | Match |
|-----|-----------|--------|-------|
| 0 | 737f767a12f54e70eecbc868... | 737f767a12f54e70eecbc868... | ✅ |
| 2 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 3 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 5 | (72 values) | 0caef3f6e685e782816fe0ce... | ✅ (in set) |
| 6 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 7 | 0215b629b95be153a9a15d7c... | 0215b629b95be153a9a15d7c... | ✅ |
| 10 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 11 | (7 values) | 857516a0408cb6ef303e5e32... | ✅ (in set) |
| 12 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 13 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 14 | b1d778bd45c6af726c0e5d78... | b1d778bd45c6af726c0e5d78... | ✅ |
| 15 | 000000000000000000000000... | 000000000000000000000000... | ✅ |

#### aarch64-aws-ecs-3-fips
- **Status**: PASS
- **Disk prediction**: PASS
- **AMI prediction**: PASS
- **AMI ID**: ami-03c94418f6a89ed0f
- **Mismatched PCRs**: None

| PCR | Predicted | Actual | Match |
|-----|-----------|--------|-------|
| 0 | 737f767a12f54e70eecbc868... | 737f767a12f54e70eecbc868... | ✅ |
| 2 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |                                                                                                 | 3 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 5 | (72 values) | 21fb8687e39e0220baa8d1e1... | ✅ (in set) |
| 6 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 7 | 0215b629b95be153a9a15d7c... | 0215b629b95be153a9a15d7c... | ✅ |
| 10 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 11 | (7 values) | 857516a0408cb6ef303e5e32... | ✅ (in set) |
| 12 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 13 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 14 | b1d778bd45c6af726c0e5d78... | b1d778bd45c6af726c0e5d78... | ✅ |
| 15 | 000000000000000000000000... | 000000000000000000000000... | ✅ |

#### x86_64-vmware-dev
- **Status**: PASS
- **Disk prediction**: PASS
- **AMI prediction**: N/A (VMware variant)
- **Mismatched PCRs**: None

| PCR | Predicted | Actual | Match |
|-----|-----------|--------|-------|
| 1 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 2 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 3 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 4 | 7f26fb1af9759301841435d7... | 7f26fb1af9759301841435d7... | ✅ |
| 6 | 3d458cfe55cc03ea1f443f15... | 3d458cfe55cc03ea1f443f15... | ✅ |
| 7 | 8a36f6d1613525fe65a7a916... | 8a36f6d1613525fe65a7a916... | ✅ |
| 9 | 331c858ffa0ef3ed08ea94c8... | 331c858ffa0ef3ed08ea94c8... | ✅ |
| 10 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 11 | (7 values) | 857516a0408cb6ef303e5e32... | ✅ (in set) |
| 12 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 13 | 000000000000000000000000... | 000000000000000000000000... | ✅ |
| 14 | b1d778bd45c6af726c0e5d78... | b1d778bd45c6af726c0e5d78... | ✅ |
| 15 | 000000000000000000000000... | 000000000000000000000000... | ✅ |



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
twoliter: predict PCR values at build time

Make `pcrsys` available in the build environment so we can generate
predictions after the disk image is assembled.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add README.md

Signed-off-by: Ben Cressey <bcressey@amazon.com>

cbgbt commented on tools/pcrsys/src/efi.rs:
```diff
@@ -0,0 +1,213 @@
+
+/// Collection of EFI variables from efi-vars.json.
+#[derive(Debug, Deserialize)]
+pub struct EfiVars {
```
Should we disallow unknown fields here?

cbgbt commented on tools/pcrsys/src/efi.rs:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    /// Validate that all required variables are present with correct GUIDs.
+    pub fn validate(&self) -> Result<()> {
```
Seems like state that we should punt to the type system. Can we do this on parse?

`#[serde(try_from = )]` is good for stuff like this if you don't want to pull in a more opinionated macro from crates.io.

cbgbt commented on tools/pcrsys/src/aws/uefi.rs:
```diff
@@ -0,0 +1,241 @@
+/// The dictionary can only be set after zlib signals it needs one.
+fn decompress_with_dict(compressed: &[u8]) -> Result<Vec<u8>> {
+    let mut decomp = Decompress::new(true);
+    let mut output = vec![0u8; 256 * 1024];
```
Mind commenting what constrains the output to actually fit into this buffer? Is it that the efivars are a known structure size?

EDIT: I think I get it now - we need to decompress to an offset so `capacity()` isn't right, but we `truncate()` later to right-size. Can we comment on how we know the output is constrained by this range?

bcressey replied:
I've replaced this with a loop that resizes the vector after the decompressor signals that it was too small. That avoids any hard-coded buffer sizes.

cbgbt reviewed (commented):
This is cool. Making my way through, will incrementally post comments. Just got the point where you actually start predicting PCRs.

cbgbt commented on tools/pcrsys/src/aws/zlib_dict.rs:
```diff
@@ -0,0 +1,233 @@
+});
+
+#[cfg(test)]
+mod tests {
```
A test that base64 decodes and compares to a checksum of the source bytes could be a good one.

cbgbt commented on tools/pcrsys/src/aws/uefi.rs:
```diff
@@ -0,0 +1,241 @@
+            let in_off =
+                usize::try_from(decomp.total_in()).whatever_context("total_in overflow")?;
+            let out_off =
+                usize::try_from(decomp.total_out()).whatever_context("total_out overflow")?;
```
Would it be wise to check that we didn't silently truncate data by comparing `total_out()` to our buffer size?

bcressey replied:
This should also be addressed by the new resize loop.

cbgbt commented on twoliter/embedded/img2img:
```diff
@@ -309,6 +309,15 @@ boot_usage=$(get_ext4_usage "${BOOT_IMAGE}" "BOOT-A")
 echo "${root_usage}" "${boot_usage}" | jq -s 'add | {disk_usage: .}' > "${OUTPUT_DIR}/artifact-metadata.json"
 
+# Predict PCR values after all signing is complete.
+if [[ "${UEFI_SECURE_BOOT}" == "yes" ]]; then
```
So basically, we unconditionally compute these. A question I am writing down (even if I just come back and answer myself is) under the circumstances where we have secure-boot variables, can we unconditionally generate useful pcr predictions?

I'm writing this comment before having read all the code, but I think the answer is partially "depends on if pcrsys supports the underlying platform."


ETA: Seemingly yes, and we have code for all of the platforms currently supported in bottlerocket-os/bottlerocket

bcressey replied:
Yup, my goal was for this to "just work" for any variant with Secure Boot support. Without Secure Boot enforced, the predictions aren't meaningful.

cbgbt commented on tools/pcrsys/README.md:
```diff
@@ -0,0 +1,86 @@
+
+## Overview
+
+`pcrsys` predicts TPM PCR values (SHA-256) for Bottlerocket images without requiring a running system. It analyzes disk images and EFI variables to compute what the PCR values would be after boot, enabling pre-registration of expected measurements for attestation policies.
```
This is extremely cool. Nice work!

cbgbt commented on tools/pcrsys/README.md:
````diff
@@ -0,0 +1,86 @@
+pcrsys disk --image /path/to/disk.img --efi-vars /path/to/efi-vars.json [--platform aws|vmware|metal]
+```
+
+### From an AWS AMI
````
I imagine this being useful and like it, but just making sure my understanding of this PR is complete - we aren't using this just yet, right?

I think a workflow I see being useful is basically unpacking an AMI to repack later. I imagine that was the incentive?

bcressey replied:
> I imagine this being useful and like it, but just making sure my understanding of this PR is complete - we aren't using this just yet, right?

What I had in mind was for using `pcrsys` as a standalone tool to work with a public Bottlerocket AMI, even if we hadn't published `pcr-predictions.json` somewhere, or you didn't know where to find it, or you just wanted to compare results.

One wrinkle is that EC2 doesn't allow DescribeImageAttribute for AMIs that you don't own, including public AMIs, so you'd first need to copy it to an AMI of your own.

cbgbt commented on tools/pcrsys/src/predict.rs:
```diff
@@ -0,0 +1,29 @@
+//! Core PCR prediction types and functions.
+
+use serde::Serialize;
+
+/// Collection of PCR predictions for JSON output.
+#[derive(Serialize)]
+pub struct PcrPredictions {
+    /// List of PCR records.
+    pub records: Vec<PcrRecord>,
+}
```
Is multiple records for the same index valid? If it's not, should we change this model to make such runtime usage impossible?

---

This is almost certainly not a good idea and perhaps unhinged, but I wonder if you could (ab)use `bon` (or something like it) to statically check that each PCR is only set once.

bcressey replied:
I've restructured the output to use a map of PCR indexes to values, and adjusted the interface for the prediction modules so they can provide both the key and the value.

cbgbt commented on tools/pcrsys/src/pcrs/pcr1.rs:
```diff
@@ -0,0 +1,38 @@
+        Platform::Vmware => Some(PcrRecord::new(1, extend_pcr_separator(&PCR_INIT_VAL))),
+        Platform::Aws | Platform::Metal => None,
+    }
+}
```
The platform-specific nature of each prediction has me wondering if it would be easier to maintain this by moving the different pcr prediction implementations into some kind of Platform trait + a factory method on CLI start that parses the input into the right one.

bcressey replied:
I tried this a few ways after refactoring the predictor interface, since the `PcrContext` struct made this more workable by streamlining the parameters.

Ultimately I felt that the logic for PCRs was sufficiently similar across platforms that I didn't see a good way to divide it into per-platform functions without either duplication or indirection into generic implementations.

cbgbt commented on tools/pcrsys/src/gpt.rs:
```diff
@@ -16,6 +17,66 @@ pub const fn uuid_to_guid(uuid: [u8; 16]) -> [u8; 16] {
+        self.0.get_bits(48..52)
+    }
+
+    /// Set priority value (bits 48-51). Panics if `priority > 15`.
```
nit: I somewhat prefer using `nutype` to validate such preconditions.

ETA: this would make it slightly easier to determine what went wrong if we validate the preconditions (e.g. error with more data, or even a panic when the faulty data is constructed rather than used.) It does complicate the API though. I'm open to a suggestion that the juice isn't worth the squeeze here.

bcressey replied:
Added a `Nibble` type via `nutype`.

cbgbt commented on tools/pcrsys/src/pcrs/pcr5.rs:
```diff
@@ -0,0 +1,205 @@
+/// Priority bit combinations: (priority, tries_left, successful)
+const COMBINATIONS: [(u8, u8, bool); 6] = [
+    (0, 0, false), // inactive
+    (0, 1, false), // inactive with tries
```
Note for posterity: Kiro suggested this may be invalid but after checking the `signpost` implementation it is decidedly valid.

cbgbt commented on tools/pcrsys/src/pcrs/pcr7.rs:
```diff
@@ -0,0 +1,323 @@
+
+    if sig_size == 0 {
+        return result;
+    }
```
Defense-in-depth question: Should we actually fail if the ESL is malformed? The value is derived from our EFI variables, which are extracted from our image.

If the data is malformed either:
* There's a serious problem with the EFI variables already
* There's a bug in our extraction code

I think failing on either case is probably better than passing, but then potentially having an attestion problem later.

bcressey replied:
Added an `EslHeader` type with more restrictive parsing.

cbgbt commented on tools/pcrsys/src/pcrs/pcr7.rs:
```diff
@@ -0,0 +1,323 @@
+    pcr = extend_pcr_separator(&pcr);
+
+    // 7. db authority - certificate used to verify shim
+    let db_sig = extract_first_sig_from_esl(db_data)?;
```
I spent some time noodling on this so leaving another posterity comment - what if there are more certs in the db? Is the signing cert always the first?

I think technically this code works because it assumes our sbkeys infrastructure which adds one cert to the db, the signing cert. That's probably fine.

```rust
// assumes single cert in db per Bottlerocket sbkeys profile
```

might be a good comment, but I wouldn't block on it.

cbgbt reviewed (commented):
Overall I found this implementation to be mechanical and straightforward to follow. The test coverage is great. Nice work!

I tend to lean more heavily on the type system and left some comments that suggest such changes, but I wouldn't block on those. Feel free to take em or leave em.

Commit:
pcrsys: add initial CLI skeleton

To do better than a "trust on first use" policy, where PCR values
from a running system are treated as the canonical values for remote
attestation, we need to generate the expected values "offline", prior
to running any code.

Prepare the skeleton for a command-line program that handles this.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add platform module

Since PCR predictions differ based on platform, add a platform module
so predictors can use the appropriate logic for the target platform.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add EFI module

Add an EFI module with structures and functions for handling EFI
variable data and EFI signature lists.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add GPT module

Add a GPT module with structures and functions for GPT labels and the
Bottlerocket-specific partition flags.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add PE module

Add a PE module with functions for working with portable executable
files.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add parsers module

Add a parsers module with functions to parse grub.cfg and bootconfig
files.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add diskfs module

Add a diskfs module with functions to extract files from exfat and
ext4 filesystems.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add AWS module

Add an AWS module with functions to find an AMI's root snapshot, and
to extract its UEFI variable data.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predict module

Add a predict module with structs and functions for generating and
serializing PCR predictions. Includes a test module with helpers for
testing predictions.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 0

AWS documents a fixed value for PCR 0:
  https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nitrotpm-attestation-document-content.html

For other platforms, this value is unpredictable.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 1

VMware firmware measures a separator value into PCR 1.

For other platforms, this value is unpredictable.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 2

PCR 2 contains firmware measurements of non-OS executables such as
UEFI drivers and Option ROM code.

For AWS and VMware, predict a separator-only value for PCR 2, i.e.
that none of these will be present at runtime.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 3

PCR 3 contains firmware measurements of configuration data for non-OS
executables, for example Option ROM configuration and data passed to
UEFI drivers.

For AWS and VMware, predict a separator-only value for PCR 3, i.e.
that none of these will be present at runtime.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 4

PCR 4 contains firmware measurements of boot loaders (shim, GRUB) and
EFI applications (linux).

Predict that shim, grub, and vmlinuz will be loaded in the expected
sequence.

Skip PCR 4 predictions when in-place updates are available, since the
kernel in either of the partition banks can change over the lifetime
of the running system.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 5

PCR 5 contains firmware measurements of the GPT header and entries.

For Bottlerocket, this is complicated by two factors. First, the
partition layout depends on whether or not in-place updates are
enabled. Second, both GRUB and the OS modify the partition table
entries at runtime, to keep track of which partition bank is active,
whether this boot was successful, and whether any boot has succeeded.

Account for both factors by enumerating all combinations of the bit
flags for the available partitions. This results in 12 predictions
for PCR 5 if in-place updates are disabled, and 72 predictions if
they're enabled.

Skip this prediction for VMware, since the firmware measures boot
variables containing random data such as MAC addresses.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 6

PCR 6 contains firmware measurements of OEM-specific attributes and
events.

For AWS and VMware, predict a separator-only value for PCR 6, i.e.
that none of these will be present at runtime.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 7

PCR 7 contains firmware measurements of state related to Secure Boot:
variables, trusted keys, SBAT policy, and the actual keys that were
used to verify the boot components.

Predict that Secure Boot will be enabled; that variables will match
their build time or AMI registration time values; that shim will be
present; that the SBAT policy will be what we expect; and that the
only machine-owner key will be the vendor key that signed the kernel.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 9

PCR 9 contains the measurement of the kernel command line.

Bottlerocket deviates from the upstream Linux kernel and measures the
final value of /proc/cmdline as seen from userspace, to ensure that
parameters set through bootconfig are also measured.

Consequently, predicting the kernel command line requires parsing
both grub.cfg for the parameters that GRUB adds, and bootconfig.data
for the parameters that the kernel adds.

Predict PCR 9 based on these two inputs.

Skip PCR 9 predictions when in-place updates are available, since the
root filesystem in either of the partition banks can change over the
lifetime of the running system, which will change the dm-verity root
hash.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 10

Predict an all-zero value for PCR 10, which is reserved for future
use.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 11

PCR 11 contains measurements of various boot milestones, indicating
whether a particular systemd target has been reached.

Provide measurements for each of the potential milestones, since any
of them can be valid depending on when the running system performs
remote attestation.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 12

Predict an all-zero value for PCR 12, which is reserved for future
use.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 13

Predict an all-zero value for PCR 13, which is reserved for future
use.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 14

PCR 14 contains measurements of shim's variables for machine-owner
keys that are allowed and denied.

Predict that only the vendor cert used to sign GRUB and the Linux
kernel will be present in MokList; that no revocation list exists in
MokListX; and that the vendor cert is trusted for module signing per
MokListTrusted.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: add predictor for PCR 15

Predict an all-zero value for PCR 15, which is reserved for future
use.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
pcrsys: predict PCRs for disk images and AMIs

For disk images, in addition to the input file, we also need certain
EFI variables in order to predict the measurements performed by the
firmware.

For AMI inputs, we retrieve the UEFI variable data, decompress it,
and parse into the standard JSON format. Then we fetch the disk image
for the root volume.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

bcressey commented:
⬆️ force push to address feedback

Re-ran tests:
| Variant | Disk vs Runtime | AMI vs Disk | PCRs Checked |
|---------|:-:|:-:|---|
| x86_64-aws-ecs-3-fips | ✅ | ✅ | 0,2,3,5,6,7,10-15 |
| aarch64-aws-ecs-3-fips | ✅ | ✅ | 0,2,3,5,6,7,10-15 |
| x86_64-vmware-dev | ✅ | N/A | 1-4,6,7,9-15 |

Notes:
- PCRs 4 and 9 skipped for AWS variants (A/B boot partitions)
- PCR 5 matched 1 of 72 predicted GPT variations on both AWS variants
- PCR 11 matched boot phase 5/7 on all variants

bcressey commented:
⬆️ force push rebase onto develop to try to address the `Cargo.lock` issue

bcressey commented:
⬆️ force push - update `Cargo.lock` to try to fix CI.

bcressey commented:
⬆️ force push - finally fixed CI.

cbgbt reviewed (approved):
lgtm, nice work! Thanks also for fixing the CI failure.

arnaldo2792 commented on tools/pcrsys/src/gpt.rs:
```diff
@@ -0,0 +1,564 @@
+    }
+
+    /// Create a minimal valid GPT for testing.
+    fn mock_gpt() -> Vec<u8> {
```
Awesome! 🎉

jmt-lab commented on tools/pcrsys/src/aws/uefi.rs:
```diff
@@ -0,0 +1,298 @@
+        .decode(b64_data.trim())
+        .whatever_context("failed to decode base64")?;
+
+    ensure_whatever!(binary.len() >= UEFI_HEADER_SIZE, "UEFI data too short");
```
nit(not-blocker): It would be nice to have proper error variants for these edge cases, I realize its a bunch of them but it can be nice. This can be a future improvement though

bcressey replied:
One of the things I like about snafu/whatever is it provides a way to level up into proper error variants if we need to, without paying for that complexity up front.

For now, since this isn't a library and there's a wide range of possible errors, collecting them into an enum felt like overkill. Happy to tackle it in the future if the usage pattern changes.

jmt-lab commented on tools/pcrsys/src/pcrs/mod.rs:
```diff
@@ -0,0 +1,17 @@
+//! PCR prediction modules for each TPM Platform Configuration Register.
```
nit(not-blocker): It would be nice if potentially we can find a way to properly trait implement these numerical prediction methods, but i'm not sure really myself the right way to go about that after looking it seems pretty straight forward this way too.
